File: api/services/directory_watcher.py
"""
Filesystem watcher service - monitors watch directories for new files
"""
import asyncio
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, Set
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler, FileCreatedEvent, FileMovedEvent
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models import WatchDirectory
from ..database import AsyncSessionLocal


# Supported image extensions
IMAGE_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.webp', '.gif', '.bmp', '.tiff', '.tif'}
VIDEO_EXTENSIONS = {'.mp4', '.webm', '.mov'}
SUPPORTED_EXTENSIONS = IMAGE_EXTENSIONS | VIDEO_EXTENSIONS


class DirectoryEventHandler(FileSystemEventHandler):
    """Handle filesystem events for a watched directory"""

    def __init__(self, directory_id: int, auto_tag: bool, loop: asyncio.AbstractEventLoop):
        super().__init__()
        self.directory_id = directory_id
        self.auto_tag = auto_tag
        self.loop = loop
        self._pending_files: Set[str] = set()
        self._debounce_delay = 1.0  # Wait 1 second for file to finish writing

    def _is_supported_file(self, path: str) -> bool:
        """Check if file has a supported extension"""
        ext = Path(path).suffix.lower()
        return ext in SUPPORTED_EXTENSIONS

    def _schedule_import(self, file_path: str):
        """Schedule file import with debouncing"""
        if file_path in self._pending_files:
            return

        self._pending_files.add(file_path)
        asyncio.run_coroutine_threadsafe(
            self._debounced_import(file_path),
            self.loop
        )

    async def _debounced_import(self, file_path: str):
        """Wait for file to finish writing, then import"""
        await asyncio.sleep(self._debounce_delay)

        # Check file still exists and is stable
        try:
            if not os.path.exists(file_path):
                self._pending_files.discard(file_path)
                return

            # Check file size is stable (not still being written)
            size1 = os.path.getsize(file_path)
            await asyncio.sleep(0.5)
            size2 = os.path.getsize(file_path)

            if size1 != size2:
                # File still being written, reschedule
                self._pending_files.discard(file_path)
                self._schedule_import(file_path)
                return

            # Import the file
            await self._import_file(file_path)

        except Exception as e:
            print(f"[Watcher] Error importing {file_path}: {e}")
        finally:
            self._pending_files.discard(file_path)

    async def _import_file(self, file_path: str):
        """Import a file into the library"""
        from .importer import import_image

        async with AsyncSessionLocal() as db:
            try:
                result = await import_image(
                    file_path,
                    db,
                    watch_directory_id=self.directory_id,
                    auto_tag=self.auto_tag
                )
                if result['status'] == 'imported':
                    print(f"[Watcher] Imported: {Path(file_path).name}")
                elif result['status'] == 'duplicate':
                    print(f"[Watcher] Duplicate: {Path(file_path).name}")
            except Exception as e:
                print(f"[Watcher] Failed to import {file_path}: {e}")

    def on_created(self, event):
        """Handle file creation events"""
        if event.is_directory:
            return
        if self._is_supported_file(event.src_path):
            self._schedule_import(event.src_path)

    def on_moved(self, event):
        """Handle file move/rename events (file moved INTO directory)"""
        if event.is_directory:
            return
        if self._is_supported_file(event.dest_path):
            self._schedule_import(event.dest_path)


class DirectoryWatcher:
    """Manages filesystem watchers for all watch directories"""

    def __init__(self):
        self.observer: Observer = None
        self.watches: Dict[int, object] = {}  # directory_id -> watch handle
        self.handlers: Dict[int, DirectoryEventHandler] = {}
        self.loop: asyncio.AbstractEventLoop = None
        self._running = False

    async def start(self):
        """Start watching all enabled directories"""
        self.loop = asyncio.get_event_loop()
        self.observer = Observer()
        self.observer.start()
        self._running = True

        # Load and watch all enabled directories
        await self._refresh_watches()
        print("[Watcher] Directory watcher started")

        # Scan for files added while app was closed
        asyncio.create_task(self._startup_scan())

    async def stop(self):
        """Stop all watchers"""
        self._running = False
        if self.observer:
            self.observer.stop()
            self.observer.join(timeout=5)
        self.watches.clear()
        self.handlers.clear()
        print("[Watcher] Directory watcher stopped")

    async def _refresh_watches(self):
        """Refresh watches from database"""
        async with AsyncSessionLocal() as db:
            result = await db.execute(
                select(WatchDirectory).where(WatchDirectory.enabled == True)
            )
            directories = result.scalars().all()

            current_ids = set(self.watches.keys())
            new_ids = {d.id for d in directories}

            # Remove watches for disabled/deleted directories
            for dir_id in current_ids - new_ids:
                self._remove_watch(dir_id)

            # Add watches for new directories
            for directory in directories:
                if directory.id not in self.watches:
                    self._add_watch(directory)

    async def _startup_scan(self):
        """Scan watched directories for files added while app was closed"""
        from .importer import import_image

        print("[Watcher] Starting startup scan for new files...")

        try:
            async with AsyncSessionLocal() as db:
                result = await db.execute(
                    select(WatchDirectory).where(WatchDirectory.enabled == True)
                )
                directories = result.scalars().all()

                total_imported = 0
                total_duplicates = 0

                for directory in directories:
                    path = Path(directory.path)
                    if not path.exists():
                        print(f"[Watcher] Skipping non-existent path: {directory.path}")
                        continue

                    # Get last scan time (or use epoch if never scanned)
                    last_scan = directory.last_scanned_at
                    if last_scan and last_scan.tzinfo is None:
                        last_scan = last_scan.replace(tzinfo=timezone.utc)

                    print(f"[Watcher] Checking {directory.name or directory.path}, last scanned: {last_scan}")

                    # Find files to scan
                    files_to_check = []
                    if directory.recursive:
                        pattern = '**/*'
                    else:
                        pattern = '*'

                    for file_path in path.glob(pattern):
                        if not file_path.is_file():
                            continue
                        if file_path.suffix.lower() not in SUPPORTED_EXTENSIONS:
                            continue

                        # Check if file was modified after last scan
                        try:
                            mtime = datetime.fromtimestamp(file_path.stat().st_mtime, tz=timezone.utc)
                            if last_scan is None or mtime > last_scan:
                                files_to_check.append(file_path)
                        except OSError:
                            continue

                    if files_to_check:
                        print(f"[Watcher] Found {len(files_to_check)} new files in {directory.name or directory.path}")

                        for file_path in files_to_check:
                            # Use a separate session for each import to avoid cascade failures
                            try:
                                async with AsyncSessionLocal() as import_db:
                                    import_result = await import_image(
                                        str(file_path),
                                        import_db,
                                        watch_directory_id=directory.id,
                                        auto_tag=directory.auto_tag
                                    )
                                    if import_result['status'] == 'imported':
                                        total_imported += 1
                                    elif import_result['status'] == 'duplicate':
                                        total_duplicates += 1
                            except Exception as e:
                                print(f"[Watcher] Error importing {file_path.name}: {e}")

                    # Update last_scanned_at
                    directory.last_scanned_at = datetime.now(timezone.utc)

                await db.commit()

            if total_imported > 0 or total_duplicates > 0:
                print(f"[Watcher] Startup scan complete: {total_imported} imported, {total_duplicates} duplicates")
            else:
                print("[Watcher] Startup scan complete: no new files found")

        except Exception as e:
            print(f"[Watcher] Startup scan FAILED: {e}")
            import traceback
            traceback.print_exc()

    def _add_watch(self, directory: WatchDirectory):
        """Add a watch for a directory"""
        path = Path(directory.path)
        if not path.exists():
            print(f"[Watcher] Skipping non-existent path: {directory.path}")
            return

        handler = DirectoryEventHandler(
            directory_id=directory.id,
            auto_tag=directory.auto_tag,
            loop=self.loop
        )

        try:
            watch = self.observer.schedule(
                handler,
                str(path),
                recursive=directory.recursive
            )
            self.watches[directory.id] = watch
            self.handlers[directory.id] = handler
            print(f"[Watcher] Watching: {directory.path} (recursive={directory.recursive})")
        except Exception as e:
            print(f"[Watcher] Failed to watch {directory.path}: {e}")

    def _remove_watch(self, directory_id: int):
        """Remove a watch for a directory"""
        if directory_id in self.watches:
            try:
                self.observer.unschedule(self.watches[directory_id])
            except Exception:
                pass
            del self.watches[directory_id]
            del self.handlers[directory_id]
            print(f"[Watcher] Stopped watching directory {directory_id}")

    async def add_directory(self, directory: WatchDirectory):
        """Add a new directory to watch"""
        if not self._running:
            return
        self._add_watch(directory)

    async def remove_directory(self, directory_id: int):
        """Stop watching a directory"""
        self._remove_watch(directory_id)

    async def refresh(self):
        """Refresh watches from database (call after directory changes)"""
        if self._running:
            await self._refresh_watches()


# Global watcher instance
directory_watcher = DirectoryWatcher()

File: src-tauri/src/services/directory_watcher.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use notify::event::{ModifyKind, RenameMode};
use notify::{Event, EventKind, RecommendedWatcher, RecursiveMode, Watcher};
use rusqlite::params;
use tokio::sync::mpsc;

use crate::server::state::AppState;
use crate::services::file_tracker;
use crate::services::importer;

/// Manages filesystem watchers for all watch directories.
pub struct DirectoryWatcher {
    state: AppState,
    /// Active watcher handles by directory_id.
    watches: Arc<Mutex<HashMap<i64, WatchHandle>>>,
    /// Shutdown signal sender.
    shutdown_tx: Option<mpsc::Sender<()>>,
}

struct WatchHandle {
    _watcher: RecommendedWatcher,
    _path: PathBuf,
    recursive: bool,
}

impl DirectoryWatcher {
    pub fn new(state: AppState) -> Self {
        Self {
            state,
            watches: Arc::new(Mutex::new(HashMap::new())),
            shutdown_tx: None,
        }
    }

    /// Start watching all enabled directories.
    pub fn start(&mut self) {
        let state = self.state.clone();
        let watches = self.watches.clone();
        let (shutdown_tx, mut shutdown_rx) = mpsc::channel::<()>(1);
        self.shutdown_tx = Some(shutdown_tx);

        tokio::spawn(async move {
            // Load enabled directories from main DB
            let directories = match load_enabled_directories(&state) {
                Ok(dirs) => dirs,
                Err(e) => {
                    log::error!("[Watcher] Failed to load directories: {}", e);
                    return;
                }
            };

            log::info!(
                "[Watcher] Starting watches for {} directories",
                directories.len()
            );

            for (dir_id, dir_path, recursive) in &directories {
                if let Err(e) = add_watch(&state, &watches, *dir_id, dir_path, *recursive) {
                    log::error!(
                        "[Watcher] Failed to watch directory {} ({}): {}",
                        dir_id,
                        dir_path,
                        e
                    );
                }
            }

            // Run startup scan for recently modified files
            for (dir_id, dir_path, recursive) in &directories {
                startup_scan(&state, *dir_id, dir_path, *recursive);
            }

            // Wait for shutdown
            shutdown_rx.recv().await;
            log::info!("[Watcher] Shutting down");
        });
    }

    /// Stop all watchers.
    pub fn stop(&mut self) {
        if let Some(tx) = self.shutdown_tx.take() {
            let _ = tx.try_send(());
        }
        if let Ok(mut w) = self.watches.lock() {
            w.clear();
        }
    }

    /// Add a watch for a single directory.
    pub fn add_directory(&self, directory_id: i64, path: &str, recursive: bool) {
        if let Err(e) = add_watch(&self.state, &self.watches, directory_id, path, recursive) {
            log::error!(
                "[Watcher] Failed to add watch for directory {}: {}",
                directory_id,
                e
            );
        }
    }

    /// Remove a watch for a directory.
    pub fn remove_directory(&self, directory_id: i64) {
        if let Ok(mut w) = self.watches.lock() {
            if w.remove(&directory_id).is_some() {
                log::info!("[Watcher] Removed watch for directory {}", directory_id);
            }
        }
    }

    /// Re-read enabled directories from the DB and sync watches.
    ///
    /// Adds watches for newly-enabled directories and removes watches for
    /// disabled/deleted ones. Call this after updating directory settings
    /// (e.g., enabling/disabling, changing recursive flag).
    pub fn refresh(&self) {
        let directories = match load_enabled_directories(&self.state) {
            Ok(dirs) => dirs,
            Err(e) => {
                log::error!("[Watcher] refresh: failed to load directories: {}", e);
                return;
            }
        };

        let new_ids: HashMap<i64, (String, bool)> = directories
            .into_iter()
            .map(|(id, path, recursive)| (id, (path, recursive)))
            .collect();

        let mut watches = match self.watches.lock() {
            Ok(w) => w,
            Err(_) => return,
        };

        // Remove watches for directories no longer enabled
        let stale_ids: Vec<i64> = watches
            .keys()
            .filter(|id| !new_ids.contains_key(id))
            .copied()
            .collect();

        for id in &stale_ids {
            if watches.remove(id).is_some() {
                log::info!("[Watcher] refresh: removed stale watch for directory {}", id);
            }
        }

        // Detect directories whose recursive flag changed — need to re-add
        let changed_ids: Vec<i64> = watches
            .iter()
            .filter_map(|(id, handle)| {
                new_ids.get(id).and_then(|(_, new_recursive)| {
                    if handle.recursive != *new_recursive {
                        Some(*id)
                    } else {
                        None
                    }
                })
            })
            .collect();

        for id in &changed_ids {
            if watches.remove(id).is_some() {
                log::info!(
                    "[Watcher] refresh: removed watch for directory {} (recursive flag changed)",
                    id
                );
            }
        }

        // Collect directories that need a watch added (new or changed)
        // (We need to drop the lock before calling add_watch since it also locks)
        let to_add: Vec<(i64, String, bool)> = new_ids
            .into_iter()
            .filter(|(id, _)| !watches.contains_key(id))
            .map(|(id, (path, recursive))| (id, path, recursive))
            .collect();

        drop(watches);

        for (dir_id, dir_path, recursive) in &to_add {
            if let Err(e) = add_watch(&self.state, &self.watches, *dir_id, dir_path, *recursive) {
                log::error!(
                    "[Watcher] refresh: failed to add watch for directory {} ({}): {}",
                    dir_id,
                    dir_path,
                    e
                );
            }
        }

        log::info!(
            "[Watcher] refresh: watches synced (removed {}, re-added {}, added {}).",
            stale_ids.len(),
            changed_ids.len(),
            to_add.len().saturating_sub(changed_ids.len())
        );
    }
}

// ─── Internal helpers ────────────────────────────────────────────────────────

fn load_enabled_directories(state: &AppState) -> Result<Vec<(i64, String, bool)>, String> {
    let conn = state
        .main_db()
        .get()
        .map_err(|e| format!("DB error: {}", e))?;

    let mut stmt = conn
        .prepare("SELECT id, path, recursive FROM watch_directories WHERE enabled = 1")
        .map_err(|e| format!("Query error: {}", e))?;

    let dirs: Vec<(i64, String, bool)> = stmt
        .query_map([], |row| {
            Ok((row.get(0)?, row.get(1)?, row.get(2)?))
        })
        .map_err(|e| format!("Query error: {}", e))?
        .filter_map(|r| r.ok())
        .collect();

    Ok(dirs)
}

fn add_watch(
    state: &AppState,
    watches: &Arc<Mutex<HashMap<i64, WatchHandle>>>,
    directory_id: i64,
    path: &str,
    recursive: bool,
) -> Result<(), String> {
    let dir_path = PathBuf::from(path);
    if !dir_path.exists() {
        return Err(format!("Directory does not exist: {}", path));
    }

    let state_clone = state.clone();
    let dir_id = directory_id;
    // Capture the tokio runtime handle so we can spawn from the notify thread
    let rt_handle = tokio::runtime::Handle::current();

    let mut watcher = notify::recommended_watcher(move |event: Result<Event, notify::Error>| {
        match event {
            Ok(ev) => handle_fs_event(&state_clone, dir_id, ev, &rt_handle),
            Err(e) => log::error!("[Watcher] Error in directory {}: {}", dir_id, e),
        }
    })
    .map_err(|e| format!("Failed to create watcher: {}", e))?;

    let mode = if recursive {
        RecursiveMode::Recursive
    } else {
        RecursiveMode::NonRecursive
    };

    watcher
        .watch(&dir_path, mode)
        .map_err(|e| format!("Failed to start watching: {}", e))?;

    // notify's inotify backend doesn't follow symlinks when recursively
    // discovering subdirectories. Walk the tree ourselves and add explicit
    // watches for any symlinked directories so their contents are monitored.
    if recursive {
        for entry in walkdir::WalkDir::new(&dir_path)
            .follow_links(true)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            if entry.path_is_symlink() && entry.file_type().is_dir() {
                if let Err(e) = watcher.watch(entry.path(), RecursiveMode::Recursive) {
                    log::warn!(
                        "[Watcher] Failed to watch symlinked dir {}: {}",
                        entry.path().display(),
                        e
                    );
                }
            }
        }
    }

    if let Ok(mut w) = watches.lock() {
        w.insert(
            directory_id,
            WatchHandle {
                _watcher: watcher,
                _path: dir_path,
                recursive,
            },
        );
    }

    log::info!(
        "[Watcher] Watching directory {} ({}) [{}]",
        directory_id,
        path,
        if recursive { "recursive" } else { "flat" }
    );

    Ok(())
}

fn handle_fs_event(state: &AppState, directory_id: i64, event: Event, rt: &tokio::runtime::Handle) {
    match event.kind {
        // New file created
        EventKind::Create(_) => {
            for path in &event.paths {
                if path.is_file() && importer::is_media_file(path) {
                    let file_path = path.to_string_lossy().to_string();
                    let state_clone = state.clone();

                    // Debounced import — wait for file to stabilize
                    rt.spawn(async move {
                        debounced_import(state_clone, directory_id, file_path).await;
                    });
                }
            }
        }

        // File removed
        EventKind::Remove(_) => {
            for path in &event.paths {
                let file_path = path.to_string_lossy().to_string();
                let state_clone = state.clone();

                rt.spawn(async move {
                    if let Err(e) = tokio::task::spawn_blocking(move || {
                        file_tracker::mark_file_missing(&state_clone, &file_path, directory_id)
                    })
                    .await
                    {
                        log::error!("[Watcher] Error marking file missing: {}", e);
                    }
                });
            }
        }

        // File moved/renamed TO this directory — treat as new file
        EventKind::Modify(ModifyKind::Name(RenameMode::To)) => {
            for path in &event.paths {
                if path.is_file() && importer::is_media_file(path) {
                    let file_path = path.to_string_lossy().to_string();
                    let state_clone = state.clone();

                    rt.spawn(async move {
                        debounced_import(state_clone, directory_id, file_path).await;
                    });
                }
            }
        }

        // File moved/renamed FROM this directory — treat as removal
        EventKind::Modify(ModifyKind::Name(RenameMode::From)) => {
            for path in &event.paths {
                let file_path = path.to_string_lossy().to_string();
                let state_clone = state.clone();

                rt.spawn(async move {
                    if let Err(e) = tokio::task::spawn_blocking(move || {
                        file_tracker::mark_file_missing(&state_clone, &file_path, directory_id)
                    })
                    .await
                    {
                        log::error!("[Watcher] Error marking moved-away file: {}", e);
                    }
                });
            }
        }

        // Other modifications — import untracked media files
        EventKind::Modify(_) => {
            for path in &event.paths {
                if path.is_file() && importer::is_media_file(path) {
                    let file_path = path.to_string_lossy().to_string();
                    let state_clone = state.clone();

                    rt.spawn(async move {
                        // Check if already tracked before importing
                        let already_tracked = tokio::task::spawn_blocking({
                            let state = state_clone.clone();
                            let fp = file_path.clone();
                            move || -> bool {
                                if let Ok(pool) = state.directory_db().get_pool(directory_id) {
                                    if let Ok(conn) = pool.get() {
                                        return conn
                                            .query_row(
                                                "SELECT COUNT(*) FROM image_files WHERE original_path = ?1",
                                                params![&fp],
                                                |row| row.get::<_, i64>(0),
                                            )
                                            .unwrap_or(0)
                                            > 0;
                                    }
                                }
                                false
                            }
                        })
                        .await
                        .unwrap_or(false);

                        if !already_tracked {
                            debounced_import(state_clone, directory_id, file_path).await;
                        }
                    });
                }
            }
        }

        _ => {}
    }
}

/// Wait for a file to stabilize (not being written to), then import it with retry.
async fn debounced_import(state: AppState, directory_id: i64, file_path: String) {
    // Wait 1 second for file to settle
    tokio::time::sleep(Duration::from_secs(1)).await;

    // Check file size stability
    let stable = tokio::task::spawn_blocking({
        let fp = file_path.clone();
        move || {
            let size1 = std::fs::metadata(&fp).map(|m| m.len()).ok();
            std::thread::sleep(Duration::from_millis(500));
            let size2 = std::fs::metadata(&fp).map(|m| m.len()).ok();
            size1.is_some() && size1 == size2
        }
    })
    .await
    .unwrap_or(false);

    if !stable {
        // File still being written, try again later
        tokio::time::sleep(Duration::from_secs(2)).await;
    }

    // Import with retry (3 attempts, 500ms backoff) for database busy errors
    let mut last_err: Option<String> = None;

    for attempt in 0..3u32 {
        let state_clone = state.clone();
        let fp = file_path.clone();

        match tokio::task::spawn_blocking(move || {
            importer::import_image(&state_clone, &fp, directory_id)
        })
        .await
        {
            Ok(Ok(result)) => {
                if result.status == importer::ImportStatus::Imported {
                    log::info!(
                        "[Watcher] Imported: {}",
                        std::path::Path::new(&file_path)
                            .file_name()
                            .and_then(|n| n.to_str())
                            .unwrap_or(&file_path)
                    );
                }
                return;
            }
            Ok(Err(e)) => {
                let msg = format!("{}", e);
                if (msg.contains("database is locked") || msg.contains("database is busy"))
                    && attempt < 2
                {
                    let backoff_ms = 500 * (attempt as u64 + 1);
                    log::warn!(
                        "[Watcher] DB busy importing {}, retrying in {}ms (attempt {}/3)",
                        file_path,
                        backoff_ms,
                        attempt + 1
                    );
                    tokio::time::sleep(Duration::from_millis(backoff_ms)).await;
                    last_err = Some(msg);
                    continue;
                }
                log::error!("[Watcher] Import error for {}: {}", file_path, e);
                return;
            }
            Err(e) => {
                log::error!("[Watcher] Task error: {}", e);
                return;
            }
        }
    }

    if let Some(err) = last_err {
        log::error!(
            "[Watcher] Import failed after 3 retries for {}: {}",
            file_path,
            err
        );
    }
}

/// Scan for files modified since the last scan time.
///
/// Respects the directory's `recursive` flag — only walks subdirectories
/// if recursive=true. Skips directories that already have a pending
/// `scan_directory` task to avoid duplicate work. Updates `last_scanned_at`
/// after completion.
fn startup_scan(state: &AppState, directory_id: i64, dir_path: &str, recursive: bool) {
    let conn = match state.main_db().get() {
        Ok(c) => c,
        Err(_) => return,
    };

    // ── Duplicate scan prevention ────────────────────────────────────────
    // Check if there's already a pending scan_directory task for this directory
    let has_pending_scan: bool = conn
        .query_row(
            "SELECT COUNT(*) FROM task_queue
             WHERE task_type = 'scan_directory'
               AND status IN ('pending', 'processing')
               AND payload LIKE ?1",
            params![format!("%\"directory_id\":{}%", directory_id)],
            |row| row.get::<_, i64>(0),
        )
        .unwrap_or(0)
        > 0;

    if has_pending_scan {
        log::info!(
            "[Watcher] Skipping startup scan for directory {} — scan task already queued",
            directory_id
        );
        return;
    }

    // Get last_scanned_at
    let last_scanned: Option<String> = conn
        .query_row(
            "SELECT last_scanned_at FROM watch_directories WHERE id = ?1",
            params![directory_id],
            |row| row.get(0),
        )
        .ok()
        .flatten();

    if last_scanned.is_none() {
        // Never scanned — skip startup scan (full scan should be triggered separately)
        return;
    }

    // Parse timestamp
    let last_dt = match last_scanned.and_then(|s| {
        chrono::NaiveDateTime::parse_from_str(&s, "%Y-%m-%dT%H:%M:%S%.f%z")
            .or_else(|_| chrono::NaiveDateTime::parse_from_str(&s, "%Y-%m-%d %H:%M:%S"))
            .ok()
    }) {
        Some(dt) => dt,
        None => return,
    };

    let dir_path_owned = dir_path.to_string();
    let dir_path = std::path::Path::new(&dir_path_owned);
    if !dir_path.exists() {
        return;
    }

    // ── Recursive flag respect ───────────────────────────────────────────
    // Only walk subdirectories if the directory has recursive=true.
    let mut new_files = Vec::new();

    let walker = walkdir::WalkDir::new(dir_path)
        .follow_links(true)
        .max_depth(if recursive { usize::MAX } else { 1 });

    for entry in walker.into_iter().filter_map(|e| e.ok()) {
        if !entry.file_type().is_file() {
            continue;
        }
        if !importer::is_media_file(entry.path()) {
            continue;
        }

        // Check if modified after last scan
        if let Ok(metadata) = entry.metadata() {
            if let Ok(modified) = metadata.modified() {
                let mod_dt = chrono::DateTime::<chrono::Utc>::from(modified).naive_utc();
                if mod_dt > last_dt {
                    new_files.push(entry.path().to_string_lossy().to_string());
                }
            }
        }
    }

    if new_files.is_empty() {
        // No new files, but still update last_scanned_at
        update_last_scanned_at(state, directory_id);
        return;
    }

    log::info!(
        "[Watcher] Startup: found {} modified files in directory {}",
        new_files.len(),
        directory_id
    );

    let state_clone = state.clone();
    tokio::spawn(async move {
        for file_path in new_files {
            let sc = state_clone.clone();
            let fp = file_path.clone();
            let _ = tokio::task::spawn_blocking(move || {
                importer::import_image(&sc, &fp, directory_id)
            })
            .await;
        }

        // ── Update last_scanned_at after scan completes ──────────────────
        update_last_scanned_at(&state_clone, directory_id);
    });
}

/// Update the `last_scanned_at` timestamp for a directory to now.
fn update_last_scanned_at(state: &AppState, directory_id: i64) {
    let now = chrono::Utc::now().to_rfc3339();
    if let Ok(conn) = state.main_db().get() {
        match conn.execute(
            "UPDATE watch_directories SET last_scanned_at = ?1 WHERE id = ?2",
            params![&now, directory_id],
        ) {
            Ok(_) => {
                log::debug!(
                    "[Watcher] Updated last_scanned_at for directory {}",
                    directory_id
                );
            }
            Err(e) => {
                log::error!(
                    "[Watcher] Failed to update last_scanned_at for directory {}: {}",
                    directory_id,
                    e
                );
            }
        }
    }
}
